### PR TITLE
[runtime env] Support python 3.10 for runtime_env conda

### LINKED ISSUE #30929: [runtime env] `conda` does not work with Python 3.10
### What happened + What you expected to happen

The `conda` field of `runtime_env` internally inserts the appropriate `ray` wheel into the conda environment.  The code to do this hasn't been updated to support Python 3.10.  Python 3.10 is currently "experimental" for Ray but we still need to fix this eventually.

```
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/utils.py", line 1264, in get_wheel_filename
    assert py_version in ["36", "37", "38", "39"], py_version
```

We should update this, and ideally add this step to the release process for new Ray Python versions (or otherwise automate this somehow)

### Versions / Dependencies

Python 3.10

### Reproduction script

```
from ray.runtime_env import RuntimeEnv
runtime_env = RuntimeEnv(
    conda={
        "dependencies":
        ["toolz", "dill", "pip", {
            "pip": ["pendulum", "ray[serve]"]
        }]
    },
    env_vars={"TF_WARNINGS": "none"}
)
ray.init(runtime_env=runtime_env)
@ray.remote
def f():
    pass

ray.get(f.remote())
```

Gives the traceback clearly indicating the error
```
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/util/client/server/proxier.py", line 703, in Datapath
    if not self.proxy_manager.start_specific_server(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/util/client/server/proxier.py", line 304, in start_specific_server
    serialized_runtime_env_context = self._create_runtime_env(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/util/client/server/proxier.py", line 253, in _create_runtime_env
    raise RuntimeError(
RuntimeError: Failed to create runtime_env for Ray client server, it is caused by:
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 342, in _create_runtime_env_with_retry
    runtime_env_context = await asyncio.wait_for(
  File "/home/ray/anaconda3/lib/python3.10/asyncio/tasks.py", line 445, in wait_for
    return fut.result()
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 297, in _setup_runtime_env
    await create_for_plugin_if_needed(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/plugin.py", line 252, in create_for_plugin_if_needed
    size_bytes = await plugin.create(uri, runtime_env, context, logger=logger)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 362, in create
    return await loop.run_in_executor(None, _create)
  File "/home/ray/anaconda3/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 340, in _create
    conda_dict = _get_conda_dict_with_ray_inserted(runtime_env, logger=logger)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 244, in _get_conda_dict_with_ray_inserted
    ray_pip = current_ray_pip_specifier(logger=logger)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/runtime_env/conda.py", line 156, in current_ray_pip_specifier
    return get_release_wheel_url()
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/utils.py", line 1308, in get_release_wheel_url
    filename = get_wheel_filename(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/utils.py", line 1264, in get_wheel_filename
    assert py_version in ["36", "37", "38", "39"], py_version
AssertionError: 310
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by architkulkarni**
Signed-off-by: Archit Kulkarni <architkulkarni@users.noreply.github.com>

## Why are these changes needed?
`conda` environments are isolated, so when `runtime_env` sets up a conda environment it must download the Ray wheel into the conda environment.  It must download the wheel that matches the current Python and Ray version running, otherwise there will be incompatibility issues between the workers that use this `runtime_env` and the other workers and Ray processes.

This PR updates the wheel name format logic to support Python 3.10.

## Related issue number
Closes #30929 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@scv119 If there is a "release process" for new Python versions, could this PR be linked there and added as a step?

We could make this more futureproof by removing the `assert py_version in ["36", "37", "38", "39", ...]`, but it would just delay the eventual breakage whenever the wheel name format changes, and we'd still need to manually update the tests.

**Review by edoakes — COMMENTED** [review]
@architkulkarni is there a way we can fail very loudly when Ray runs a new python version but it isn't supported in this enum? At a minimum we can add a small CI test (just check current py version vs. this map)

**Comment by edoakes** [comment]
Or maybe the existing test is good enough... just checking we catch this as early as possible

**Comment by architkulkarni** [comment]
@edoakes Good point, the existing tests will just pass even if a new version is added.

I added a constant `SUPPORTED_PY_VERSIONS` and a unit test to check the current py version against it to make sure the constant stays updated.

Unfortunately, there's still no way to auto-update the `test_get_*_wheel` CI tests because of a chicken and egg situation. The test needs wheels to be uploaded to AWS in the standard location, but the wheels get uploaded there only after the PR is merged. It's probably not too bad to have a manual process here because new python versions only come once a year.

File: python/ray/_private/utils.py [hunk 1]
```diff
@@ -1256,28 +1256,32 @@ def get_wheel_filename(
...
-    assert py_version in ["36", "37", "38", "39"], py_version
+    assert py_version in ray_constants.SUPPORTED_PY_VERSIONS, py_version
 
+    py_version_str = "".join(map(str, py_version))
+    if py_version_str in ["36", "37"]:
+        darwin_os_string = "macosx_10_15_intel"
+    elif py_version_str in ["38", "39"]:
+        darwin_os_string = "macosx_10_15_x86_64"
+    else:
+        darwin_os_string = "macosx_10_15_universal2"
```
Comment (architkulkarni) [inline python/ray/_private/utils.py:1273]: New OS string just dropped.

Comment (edoakes) [reply python/ray/_private/utils.py:1273]: lol

**Comment by architkulkarni** [comment]
`test_huggingface` was broken on master, unrelated
rllbi `test_memory_leak` was flaky on master, unrelated

**Comment by architkulkarni** [comment]
@scv119 @edoakes Any other comments on this PR?

File: python/ray/_private/utils.py [hunk 2]
```diff
@@ -1304,7 +1308,7 @@ def get_release_wheel_url(
     ray_commit: str = ray.__commit__,
     sys_platform: str = sys.platform,
     ray_version: str = ray.__version__,
-    py_version: str = f"{sys.version_info.major}{sys.version_info.minor}",
+    py_version: Tuple[int, int] = sys.version_info[:2],
```
Comment (rkooo567) [inline python/ray/_private/utils.py:1311]: why do we fix this? The previous code seems to be easier to understand

File: python/ray/tests/test_runtime_env.py [hunk 3]
```diff
@@ -37,13 +37,16 @@
...
+
 
 def test_get_wheel_filename():
+    """Test the code that generates the filenames of the `latest` wheels."""
     # NOTE: These should not be changed for releases.
     ray_version = "3.0.0.dev0"
     for sys_platform in ["darwin", "linux", "win32"]:
-        for py_version in ["36", "37", "38", "39"]:
-            if sys_platform == "win32" and py_version == "36":
+        for py_version in [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]:
```
Comment (rkooo567) [inline python/ray/tests/test_runtime_env.py:48]: Use the same constant?

**Review by rkooo567 — APPROVED** [review]
minor comments

File: python/ray/_private/utils.py (hunk 2 above)
Comment (architkulkarni) [reply python/ray/_private/utils.py:1311]: No strong reason, I just think it's better not to convert it to a string until the very end of the data flow, when we actually generate the wheel name.

But I agree that `sys.version_info[:2]` is harder to understand. I'll change it to `(sys.version_info.major, sys.version_info.minor)`

**Pull request merged**